Core/include/Acts/TrackFinding/MeasurementSelector.hpp

```diff
@@ -106,60 +107,85 @@ class MeasurementSelector {
       return CombinatorialKalmanFilterError::MeasurementSelectionFailed;
     }
 
+    assert(!cuts->chi2CutOff.empty());
+    const auto& chi2CutOff = cuts->chi2CutOff;
+    auto maxChi2Cut = *std::max_element(chi2CutOff.begin(), chi2CutOff.end());
     double minChi2 = std::numeric_limits<double>::max();
     size_t minIndex = 0;
-    size_t index = 0;
-    // Loop over all measurements to select the compatible measurements
-    for (auto& trackState : candidates) {
-      // Take the parameter covariance
-      // const auto predicted = tackState.predicted();
-      // const auto predictedCovariance = trackState.predictedCovariance();
-
-      double chi2 = calculateChi2(
-          // This abuses an incorrectly sized vector / matrix to access the
-          // data pointer! This works (don't use the matrix as is!), but be
-          // careful!
-          trackState
-              .template calibrated<MultiTrajectoryTraits::MeasurementSizeMax>()
-              .data(),
-          trackState
-              .template calibratedCovariance<
-                  MultiTrajectoryTraits::MeasurementSizeMax>()
-              .data(),
-          trackState.predicted(), trackState.predictedCovariance(),
-          trackState.projector(), trackState.calibratedSize());
-
-      trackState.chi2() = chi2;
-
-      // Search for the measurement with the min chi2
-      if (chi2 < minChi2) {
-        minChi2 = chi2;
-        minIndex = index;
+    auto trackStateIterEnd = candidates.end();
+    {
+      auto trackStateIter = candidates.begin();
+      // Loop over all measurements to select the compatible measurements
+      // Sort track states which do not satisfy the chi2 cut to the end.
+      // When done trackStateIterEnd will point to the first element that
+      // does not satisfy the chi2 cut.
+      assert(trackStateIter != trackStateIterEnd);
```

Comment:
this will always be `false` no? since we check `candidates.empty()` above
```suggestion
```

Reply:
- Yes unless the code is altered. It is just there to state the intention.